### PR TITLE
added methods,loss and bashscript for meanshift algorithm

### PR DESCRIPTION
**Pull request by ankitpatnala**
Hi, 

I am trying to implement this paper [https://arxiv.org/abs/2105.07269]() which was ICCV accepted paper. Though i couldn't able to run proper tests for that (hyperparameter tuning). It will be nice if you can comment on this or look forward to implement in your codebase

### CODE REVIEW

**Comment by vturrisi** [comment]
Hey, thanks for contributing!
I'll put it in my queue and soon I'll take a look.

**Comment by ankitpatnala** [comment]
Ya ! i ran it for CIFAR10 using the hyperparameters mentioned in bash_scripts but the results were not so great. I guess we need to tune hyperparameters and repeat experiments with hyperparameters mentioned in paper. Will you able to share plots for BYOL ? ( the difference with BYOL and mean-shift is that mean-shift is assymetric and has decrease the space to multiple neighbors)

**Comment by ankitpatnala** [comment]
Hi @vturrisi Have you ever tried simply asymmetric BYOL ? How is the performance with that ?

**Comment by vturrisi** [comment]
Hey, sorry about the delay. I'll share the plots today with you, but if the results are not good on cifar10, I would say there's something wrong with the implementation. Try also cifar100, since training time practically the same and it provides a much more difficult scenario. Apart from that, please pull from the main as we did multiple changes in the last few weeks, add tests and the method to the documentation so that we can eventually merge when it produces good results.

**Comment by ankitpatnala** [comment]
> Hey, sorry about the delay. I'll share the plots today with you, but if the results are not good on cifar10, I would say there's something wrong with the implementation. Try also cifar100, since training time practically the same and it provides a much more difficult scenario. Apart from that, please pull from the main as we did multiple changes in the last few weeks, add tests and the method to the documentation so that we can eventually merge when it produces good results.

Hi @vturrisi  sorry for the delay ! I am on holidays, so i am not so active now a days. I found an error in implementation. I will implement and let you know !

File: solo/losses/meanshiftloss.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+    Returns:
+        torch.Tensor: Mean shift loss.
+    """
+    p = F.normalize(p,dim=1)
+    z = F.normalize(z,dim=1)
+    queue = F.normalize(queue,dim=1)
+    _,indices = torch.topk(z@queue.T,num_neighbors,dim=1)
+    nn_targets = queue[indices]
+    z = z.unsqueeze(dim=1)
```
Comment (ankitpatnala) [inline solo/losses/meanshiftloss.py:24]: This should be p not z

**Comment by vturrisi** [comment]
@ankitpatnala no worries, take your time.

**Comment by ankitpatnala** [comment]
I have made few commits and now results are better. I can share the plots and results.

**Comment by vturrisi** [comment]
@ankitpatnala what did you get? About running on slurm, we just create an auxiliary bash script for slurm that calls our normal bash scripts, so the main file shouldn't be changed. We now missing the docs and tests.

**Comment by ankitpatnala** [comment]
Hi @vturrisi , I got an accuracy of 91.36 val_1 on CIFAR10 data

**Comment by vturrisi** [comment]
@ankitpatnala could you run CIFAR-100 as well?

**Comment by ankitpatnala** [comment]
> @ankitpatnala could you run CIFAR-100 as well?

I don't have CIFAR100 data in my cluster.
can you able to see this plots https://wandb.ai/a-patnala/solo-learn?workspace=user-a-patnala ?

**Comment by vturrisi** [comment]
I can't.

**Comment by ankitpatnala** [comment]
Can you share me your email registered with wandb ?

**Comment by vturrisi** [comment]
Yes, now it's fine. Plots look ok, but cifar10 results can be misleading. Can you download cifar100 an test there? Training time will be the same.

**Comment by vturrisi** [comment]
Btw, batch size is per gpu, so you are using a batch of 1024 instead of 256 that we used for the other models.

**Comment by ankitpatnala** [comment]
> Btw, batch size is per gpu, so you are using a batch of 1024 instead of 256 that we used for the other models.

Yes ! i used the same config for BYOL (assymetric and symmetrric) and BYOL is insensitive to batchsize, so i guess scores will slightly alter.

**Comment by ankitpatnala** [comment]
> Yes, now it's fine. Plots look ok, but cifar10 results can be misleading. Can you download cifar100 an test there? Training time will be the same.

Yes ! I am running it on 4 GPUs. You can use the same link to see the plots !

**Comment by ankitpatnala** [comment]
@vturrisi did you check ? If yes, is that okay ?

**Comment by ankitpatnala** [comment]
> @vturrisi did you check ? If yes, is that okay ?

@vturrisi did you check the plots ?

**Comment by vturrisi** [comment]
Hey @ankitpatnala , I'm on vacation so I'll only be able to properly check when I'm back. What is still missing is reverting back the `main_pretrain.py` file, pulling from the main branch, documentations, and tests.

**Comment by ankitpatnala** [comment]
> Hey @ankitpatnala , I'm on vacation so I'll only be able to properly check when I'm back. What is still missing is reverting back the `main_pretrain.py` file, pulling from the main branch, documentations, and tests.

Hey @vturrisi  I did all the changes and I have already pulled your main branch before but still before reviewing let me know, I will pull again if there are any new implemented features in between

**Comment by vturrisi** [comment]
@ankitpatnala I've fixed some stuff. I'll run on imagenet100 until the end of the week and then I'll get back to you

**Comment by ankitpatnala** [comment]
Hi @vturrisi  any updates ?

**Comment by vturrisi** [comment]
hi @ankitpatnala, I tried running on Imagenet-100 a couple of times but performance is suboptimal. We are deciding if we could merge this as an experimental method or if we will wait until we get good results there so that it complies with the rest of the methods.

**Comment by vturrisi** [comment]
Hi @ankitpatnala. I'll be closing this PR until we have good performance on larger scale datasets (at least Imagenet-100). Also, the library changed quite a bit and merging the main branch into this one is needed. Feel free to reopen if you manage to do both, otherwise it will be in my pipeline, but I'm very busy atm so sorry about that.
